File: app/src/main/java/com/example/cashflow/MainActivity.java
package com.example.cashflow;

import android.content.pm.PackageManager;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.example.cashflow.dataClass.*;
import com.example.cashflow.db.*;

import java.text.SimpleDateFormat;

import java.util.ArrayList;
import java.util.Calendar;

import android.Manifest;

public class MainActivity extends AppCompatActivity {
    private Button btnHome;
    private ArrayList<Account> accounts;

    private SQLiteDB sqLiteDB;
    private readSQL readSQL;
    private writeSQL writeSQL;
    private static final int PERMISSION_REQUEST_CODE = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        sqLiteDB = new SQLiteDB(this);
        SQLiteDatabase db = sqLiteDB.getWritableDatabase();

        sqLiteDB.onCreate(db);

//        sqLiteDB.clearAllTableData(db);
        readSQL = new readSQL(db);
        writeSQL = new writeSQL(db);

        accounts = readSQL.getAllAccounts();

        //accounts = null;
        System.out.println(accounts);
        if (accounts == null || accounts.size() == 0) {
            System.out.println("accounts is null");
            // Le righe di codice devono essere eseguite solo all'installazione dell'app.
            writeSQL.createAccount("Cash", 0);
            writeSQL.createAccount("Bank", 0);

            writeSQL.createCategory("Salary", "stipendio");

            City city = new City();
            writeSQL.createCity(city.getNameCity(), city.getLatitude(), city.getLongitude());

            writeSQL.createTransaction(1, 100, "2021-01-01", 1, 1, 1);

            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
            String date = sdf.format(Calendar.getInstance().getTime());
            writeSQL.createTransaction(1, 100, date, null, 1, 1);
        }

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            // Permission is not granted, request it
            requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, PERMISSION_REQUEST_CODE);
        }

        btnHome = findViewById(R.id.btnHome);
        btnHome.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loadFragment(new HomeFragment(null));
            }
        });
        loadFragment(new HomeFragment(null));

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                loadFragment(new HomeFragment(accounts));
            } else {
                System.out.println("Location permission is required to fetch the location");
            }
        }
    }

    private void loadFragment(Fragment fragment) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.fragment_container, fragment);
        fragmentTransaction.addToBackStack(null);
        fragmentTransaction.commit();
    }
}

File: app/src/main/java/com/example/cashflow/MainActivity.kt
package com.example.cashflow

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.viewModels
import com.google.android.material.navigation.NavigationView
import com.example.cashflow.box.*
import com.example.cashflow.dataClass.*
import com.example.cashflow.db.*
import com.example.cashflow.utils.Posizione


class MainActivity : AppCompatActivity() {
    private val viewModel: DataViewModel by viewModels()
    private var readSQL: ReadSQL? = null
    private var writeSQL: WriteSQL? = null


    private var accounts: ArrayList<Account>? = null
    private var city: City? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val drawerLayout: DrawerLayout = findViewById(R.id.drawer_layout)
        val navigationView: NavigationView = findViewById(R.id.nav_view)
        val myToolbar = findViewById<Toolbar>(R.id.my_toolbar)
        val hamburgerMenu: ImageButton = findViewById(R.id.menu_hamburger)
        val btnHome: ImageView = findViewById(R.id.logo)
        val toolbar_title: TextView = findViewById(R.id.toolbar_title)

        var posizione: Posizione? = null

        readSQL = viewModel.getReadSQL()
        writeSQL = viewModel.getWriteSQL()



//        val isDeleted: Boolean = this.deleteDatabase("cashflow.db")
//        if (isDeleted) {
//            Log.d("SQLiteDB", "Database deleted successfully")
//        } else {
//            Log.d("SQLiteDB", "Failed to delete database")
//        }

        accounts = readSQL!!.getAccounts()

        if (ContextCompat.checkSelfPermission(
                this,
                android.Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(android.Manifest.permission.ACCESS_FINE_LOCATION),
                PERMISSION_REQUEST_CODE
            )
        }
        posizione = Posizione(this)
        posizione!!.requestDeviceLocation(object : Posizione.DeviceLocationCallback {
            override fun onLocationFetched(city: City?) {
                this@MainActivity.city = city ?: return
            }

            override fun onLocationFetchFailed(e: Exception?) {
                Log.e("MainActivity", "Failed to fetch location", e)
            }
        })

        setSupportActionBar(myToolbar)
        supportActionBar?.setDisplayShowTitleEnabled(false)
        navigationView.setNavigationItemSelectedListener { menuItem ->
            city?.let {
                loadMenuManagerFragment(menuItem.itemId, it)
                Log.d("MainActivity", "City: $it")
            } ?: run {
                loadMenuManagerFragment(menuItem.itemId, city!!)
                Log.e("MainActivity", "City: $city")
            }
            drawerLayout.closeDrawer(GravityCompat.START)
            true
        }

        hamburgerMenu.setOnClickListener {
            drawerLayout.openDrawer(GravityCompat.START)
        }

        myToolbar.setNavigationOnClickListener {
            drawerLayout.closeDrawer(GravityCompat.START)
            true
        }

        btnHome.setOnClickListener {
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)
            null
        }
        toolbar_title.setOnClickListener {
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)
            null
        }
        loadBoxManagerFragment()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                loadBoxManagerFragment()
            } else {
                println("Location permission is required to fetch the location")
            }
        }
    }

    private fun loadBoxManagerFragment() {
        val fragment = box_manager_fragment()
        supportFragmentManager.beginTransaction().apply {
            replace(R.id.linearContainer, fragment)
            commit()
        }
    }

    private fun loadMenuManagerFragment(selectedMenuId: Int, city: City) {
        val menuManagerFragment = MenuManagerFragment.newInstance(selectedMenuId, city)
        supportFragmentManager.beginTransaction().apply {
            replace(R.id.linearContainer, menuManagerFragment)
            addToBackStack(null)
            commit()
        }
    }


    companion object {
        private const val PERMISSION_REQUEST_CODE = 1
    }
}
